text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,451 @@
+
+This will make sure that old CLI functionality is still working, assuming of course it was covered by our integration tests.
+
+# Detailed Design
+
+There are 2 mechanisms we need to implement:
+
+## API Compatibility Checks
+
+We need to make sure the objects that make up what we refer to as the `cx-protocol`, do not change in a breaking way.
+We should be able to leverage `jsii` to accomplish this. For example, this is an excerpt from the `.jsii` spec of the `cx-api` module:
+
+```json
+"name": "ContainerImageAssetMetadataEntry",
+"properties": [
+    {
+        "abstract": true,
+        "docs": {
+            "stability": "experimental",
+            "summary": "Path on disk to the asset."
+        },
+        "immutable": true,
+        "locationInModule": {
+            "filename": "lib/assets.ts",
+            "line": 48
+        },
+        "name": "path",
+        "type": {
+            "primitive": "string"
+        }
+    }
+]
+```
+
+This means that, if we remove the `path` property from `ContainerImageAssetMetadataEntry`, jsii should produce an error, and indeed:
+
+```console
+❯ yarn compat
+PROP @aws-cdk/cx-api.ContainerImageAssetMetadataEntry.path: has been removed [removed:@aws-cdk/cx-api.ContainerImageAssetMetadataEntry.path]
+```
+
+Currently, the `cx-api` module contains a lot of stuff that aren't strictly related to the protocol.
+
+> For example, `AssemblyBuildOptions`, which is used during synthesis by the framework. It is not used by the CLI (nor it should).
+
+This creates clutter, and might also cause issues with applying `jsii` compatibility checks.
+
+In addition, according to our new compatibility model, we need to make sure the user gets the correct error
+when the CLI version is smaller than the `Cloud-Assembly` version.
+
+The plan therefore is as follows:
+
+### Step 1: Create a separate package called `cx-protocol`
+
+This package will provide:
+
+#### (1) Object Model
+
+- CloudAssembly
+- RuntimeInfo
+- AssemblyManifest
+- CloudArtifact
+- ...
+- ...
+
+These are the objects that will be later used by the CLI. Running compatibility checks on them **should** ensure:
+
+1. Rename/Remove properties is not allowed.
+2. Changing the type of a property is not allowed.
+3. Making an optional property required is not allowed.
+4. Adding a required property is not allowed.
+
+> The `jsii` compatibility checker currently doesn't actually provide us with all those requirements. It validates interfaces only with relation to their
+> usage, in either argument or return values of methods. This means that, for example, number 4 is allowed,
+> because the `ContainerImageAssetMetadataEntry` is not used as a function argument in `cx-api`.
+> In order to achieve this, we will either need to enhance the `jsii` checker, or refactor our code.
+In any case, we stipulate the necessary changes will be done.
```

Comment:
I understand the reasoning behind `jsii-diff` that makes it skip over requirement number 4 for structs. 
If we take `ContainerImageAssetMetadataEntry` as an example, in the `.jsii` spec of `cx-api`, this struct isn't used as an argument to any function.

I guess `jsii` assumes the way this struct is used by users is:

```typescript
const metadata = someFunction.returnAssetMetadata()
```

In this case, adding a required property really doesn't break the API as far as the user is concerned, because he is not the one initializing it. 

However, this neglects the case where a user simply does:

```typescript
const metadata: ContainerImageAssetMetadataEntry = {}
```
In this case, adding a required property breaks the users ability to initialize the struct. 

Perhaps, we should add ability to `jsii-diff` that basically (optionally) treats struct initializers as regular methods. 

Thoughts?

Reply:
That sounds weird to me, and I reckon `jsii-diff` should disallow adding required fields to structs (or anywhere else, really), because otherwise it does not really provide the value it claims to provide. Whether the struct is used as an argument or return value only changes the severity of the breakage it would cause...

Reply:
I half agree. I mean it definitely struck me as odd, which is why i started investigating. The only scenario I can think of where this behavior is useful is when the `struct` itself isn't exported, or to more accurate, can't be directly instantiated by the user.

In java for example you can use a private constructor and factory methods, the API is then validated against those factory methods, and in this case its ok to add a required field. In typescript (and go) - you can return an unexported value from an exported function, which also makes it ok.

But from your comment i gather that was not the original intent? Should I just create a task in jsii to fix this?

Reply:
As mentioned about, the CloudAssembly interface will need to be placed in an "input position" in some other API in order to be checked for compatibility like we want it.

Reply:
I'm afraid it's not so simple. Actually, there is already a method in `cx-api` that uses `CloudAssembly` as an input:

```typescript
export class CloudArtifact {
  public static fromManifest(assembly: CloudAssembly, id: string, artifact: ArtifactManifest): CloudArtifact | undefined {}
}
``` 
This still doesn't prevent us from adding a required property to `AssemblyManifest`, which is the type of the `manifest` property in `CloudAssembly`. 
We would have to make sure all transitive objects are also used as an input somewhere.

Btw, I think adding required arguments to the `CloudAssembly` constructor is not allowed even without it being used as an input. This is because it's kind is `class`, and it has an `initializer` in the `.jsii` spec.  So adding a required argument would require more from the caller of this initializer. I that sense, i think we should be treating structs as classes that have an initializer that accepts all the properties. This initializer is essentially whats being called by `JSON.parse` anyway. 


This way we don't need any dummy methods and we don't need to worry we missed some transitive object.

Reply:
The new @aws-cdk/cloud-assembly-api (or -spec) should include only data structures (in jsii these are represented as interfaces that don’t have an “I” prefix and include only read only properties. These structs should basically represent only the schema for the cloud assembly manifest, and can be enforced via jsii-diff by including some dummy static method that uses them as inputs.

Reply:
Yeah I agree. Im just saying that instead of adding this dummy method, we can enhance `jsii` to "auto-generate" it if requested. Otherwise, we are in risk of introducing new data structures and forgetting to update the dummy method. Let's see if we can avoid having to remember stuff.

Reply:
Ideally (not 100% sure it's possible) we should just have a single root interface (`CloudAssembly`) that everything falls off from.

Reply:
After an online sync we settled on having a single dummy function that accepts the root interface as an input. Assuming this validation is done recursively (need to double check), this should be sufficient and wouldn't require us to "remember" anything.

Reply:
BTW, it doesn’t need to be a dummy function, we can do something like:

```ts
saveManifest(manifest: CloudAssembly): void
```